Implement Mayhem

nottirb opened the pull request:
- Added a `Dockerfile` for fuzz testing
- Added `Mayhemfile(s)` for fuzz testing
- Added a `mayhem.yml` file for automatically running fuzz tests with Mayhem using the mcode action
- Wrote 2 fuzz harnesses
    - One allows for much faster fuzzing and is decent at finding new defects
    - One allows for much slower fuzzing and is much better at analyzing all paths for a test input

nottirb committed:
Implement Mayhem

nottirb committed:
Implement Mayhem

nottirb committed:
Fix Errors

nottirb committed:
Fix Errors

nottirb committed:
Add Image Tag

nottirb committed:
Add Image Tag

nottirb committed:
Update fuzz_process_rand.rs

nottirb committed:
Merge branch 'main' of https://github.com/nottirb/cargo-deb

nottirb committed:
Fix Some Fuzzing Setup Issues

nottirb committed:
Bump Fuzz Duration to 30

xansec commented on fuzz/fuzz_targets/fuzz_process_deep.rs:
```diff
@@ -0,0 +1,240 @@
+use honggfuzz::fuzz;
```
Performance may degrade for sufficiently large input indices. I recommend putting a cap on the for loop from 0..=input.len() to something fixed.

xansec commented on fuzz/fuzz_targets/fuzz_process_rand.rs:
```diff
@@ -0,0 +1,257 @@
+use honggfuzz::fuzz;
```
Same comment as previous fuzz target. Also looking for confirmation that the seeded random is guaranteed to be fixed.

xansec reviewed (changes requested):
The testsuites are also a bit large, I wouldn't commit more than 10-20 in general. If you feel it is necessary we can keep. Please make the requested changes.

nottirb committed:
Swap out StdRng for Pcg64

nottirb committed:
Remove Images

nottirb committed:
Merge branch 'main' of https://github.com/nottirb/cargo-deb

nottirb committed:
Limit split iteration to 32

nottirb committed:
Reduce Test Suite Size

nottirb committed:
Unify Test Suite

nottirb committed:
Limit deep splits to 100

nottirb commented:
I removed the image name from the Mayhemfiles, and replaced StdRng with Pcg64 to ensure that it generates the same random number on different builds/platforms.

Also modified the deep process fuzzer to use a max of 100 split indices (randomized). Signficantly reduced the testsuite size.

Workflow might error (github packages is down as of writing this comment). I verified everything was working using dockerhub for the image with a manual mayhem run. Can re-run jobs if needed.

xansec reviewed (approved):
LGTM!
